src/GravityVisitor.h

```diff
@@ -9,48 +9,41 @@ extern CProxy_Resumer<CentroidData> centroid_resumer;
 
 class GravityVisitor {
 private:
-  const Real gconst = 1; // Per Tom's suggestion
-  const Real theta = 0.7;
-  void addGravityLeaf(const SpatialNode<CentroidData>& source, SpatialNode<CentroidData>& target) {
+  // note gconst = 1
+  static constexpr Real theta = 0.7;
+  static constexpr int  nMinParticleNode = 6;
+  void addGravity(const SpatialNode<CentroidData>& source, SpatialNode<CentroidData>& target) {
     int curr_counter = 0;
     for (int i = 0; i < target.n_particles; i++) {
-      Vector3D<Real> sum_force;
-      for (int j = 0; j < source.n_particles; j++) {
-        if (target.particles[i].key == source.particles[j].key) continue;
-        curr_counter++;
-        Vector3D<Real> diff = source.particles[j].position - target.particles[i].position;
-        Real rsq = diff.lengthSquared();
-        sum_force += diff * (source.particles[j].mass / (rsq * sqrt(rsq)));
-      }
-      target.applyForce(i, gconst * target.particles[i].mass * sum_force);
-    }
-#if COUNT_INTRNS
-    centroid_resumer.ckLocalBranch()->countInts(curr_counter);
-#endif
-  }
-  void addGravityNode(const SpatialNode<CentroidData>& source, SpatialNode<CentroidData>& target) {
-    for (int i = 0; i < target.n_particles; i++) {
-      Vector3D<Real> diff = source.data.centroid - target.particles[i].position;
+      Vector3D<Real> diff = source.data.centroid - target.particles()[i].position;
       Real rsq = diff.lengthSquared();
-      Real scalar = (gconst * source.data.sum_mass * target.particles[i].mass / (rsq * sqrt(rsq)));
-      target.applyForce(i, diff * scalar);
+      if (rsq != 0) {
+        Vector3D<Real> accel = diff * (source.data.sum_mass / (rsq * sqrt(rsq)));
+        target.applyAcceleration(i, accel);
+      }
     }
-#if COUNT_INTRNS
-    centroid_resumer.ckLocalBranch()->countInts(-target.n_particles);
-#endif
   }
   public:
   GravityVisitor() {}
   void leaf(const SpatialNode<CentroidData>& source, SpatialNode<CentroidData>& target) {
-    addGravityLeaf(source, target);
+    addGravity(source, target);
+#if COUNT_INTRNS
+    centroid_resumer.ckLocalBranch()->countInts(target.n_particles);
+#endif
   }
   bool node(const SpatialNode<CentroidData>& source, SpatialNode<CentroidData>& target) {
+    if (source.n_particles <= nMinParticleNode) return true;
     Vector3D<Real> dr = source.data.centroid - target.data.centroid;
     Real dsq = dr.lengthSquared();
     if (theta * dsq < source.data.rsq) {
       return true;
     }
-    if (source.data.sum_mass > 0) addGravityNode(source, target);
+    if (source.data.sum_mass > 0) {
+      addGravity(source, target);
+#if COUNT_INTRNS
```

Comment:
May be documented elsewhere, but what does the `COUNT_INTRNS` macro represent?

Reply:
count interactions. the driver will receive the finished reduction sum and print it out at destroy time.

Reply:
Ah, I see. We may consider refactoring it to `COUNT_INTERACTIONS` in the future, as it's not that much longer and more self-explanatory. Regardless, this was a somewhat unrelated tangent.